Feature/ae 2036 aineisto s3 endpoint

solita-juhohaa opened the pull request.

solita-juhohaa committed:
AE-2036: Add access to MinIO Console in local development environment

solita-juhohaa committed:
AE-2036: Add `aineisto` system-user

solita-juhohaa committed:
AE-2036: Add middleware to wrap request with `aineisto` system-user

solita-juhohaa committed:
AE-2036: Add functions to call `aws-api` for multipart uploads

solita-juhohaa committed:
AE-2036: Add function to upsert file in parts

solita-juhohaa committed:
AE-2036: Refactor useful functions from api to service level

solita-juhohaa committed:
AE-2036: Add functions to update aineistot in S3

solita-juhohaa committed:
AE-2036: Add endpoint to update aineistot

solita-juhohaa committed:
AE-2036: Test that generation of aineistot returns correct response

solita-juhohaa committed:
AE-2036: Test that aineistot are actually generated

Juholei commented on docker/docker-compose.yml:
```diff
@@ -44,11 +44,12 @@ services:
     volumes:
       - data:/files 
-    command: server /files
+    command: server /files --console-address ":9001"
```
Voisi lisätä readmehen ohjeen miten tänne pääsee (ja mikä tämä on)

solita-juhohaa commented on etp-backend/src/main/clj/solita/etp/service/aineisto.clj:
```diff
@@ -65,3 +65,28 @@
+  {:banks          energiatodistus-csv/energiatodistukset-bank-csv
+   :tilastokeskus  energiatodistus-csv/energiatodistukset-tilastokeskus-csv
+   :anonymized-set energiatodistus-csv/energiatodistukset-anonymized-csv})
+
```
Ei varmaankaan huutomerkkiä kuulu olla

solita-antti-mottonen commented on etp-backend/src/main/clj/solita/etp/service/aineisto.clj:
```diff
@@ -65,3 +66,50 @@
+        ;; is the minimum requirement by `upload-part-fn`.
+        current-part (java.nio.ByteBuffer/allocate (* 8 1024 1024))
+        upload-parts-fn (fn [upload-part-fn] (csv-reducible-query (fn [row]
+                                                                    (let [row-bytes (.getBytes row "UTF-8")]
```
Tähän saa varmaan tilalle StandardCharsets/UTF_8?

solita-antti-mottonen replied:
Löytyypi täältä `(:import (java.nio.charset StandardCharsets))`

solita-juhohaa replied:
Tuo kohta nyt hukkui, kun force pushasin, mutta muutos tehty.

solita-juhohaa commented on etp-backend/src/main/clj/solita/etp/api/aineisto.clj:
```diff
@@ -72,3 +75,19 @@
+    ["/update"
+     {:post {:summary    "Päivitä kaikkien aineistojen CSV-tiedostot S3:ssa."
+             :middleware [[security/wrap-whoami-for-internal-aineisto-api]]
+             :access     rooli-service/paakayttaja?
```
Tämä on ylimääräistä

solita-juhohaa commented on etp-backend/src/test/clj/solita/etp/service/aineisto_test.clj:
```diff
@@ -74,3 +74,21 @@
+(t/deftest update-aineistot-test
+  (let []
+    (t/testing "Aineistot don't exist before generating"
+      (t/is (not (solita.etp.service.file/file-exists? ts/*aws-s3-client* "/aineistot/1/energiatodistukset.csv")))
```
not -> false?

solita-juhohaa commented on etp-backend/src/test/clj/solita/etp/service/aineisto_test.clj:
```diff
@@ -74,3 +74,21 @@
+      (t/is (not (solita.etp.service.file/file-exists? ts/*aws-s3-client* "/aineistot/3/energiatodistukset.csv"))))
+    (t/testing "Aineistot exist after generating"
+      (aineisto/update-aineistot-in-s3! ts/*db* {:id -5 :rooli 2} ts/*aws-s3-client*)
+      (t/is (solita.etp.service.file/file-exists? ts/*aws-s3-client* "/aineistot/1/energiatodistukset.csv"))
```
-> true?

solita-juhohaa replied:
require alias

solita-juhohaa commented on etp-backend/src/test/clj/solita/etp/service/aineisto_test.clj:
```diff
@@ -74,3 +74,21 @@
     (t/is (true? (aineisto/check-access ts/*db*, user-id-with-allowed-ip, allowed-aineisto-type, actual-ip)))))
+
+(t/deftest update-aineistot-test
+  (let []
```
tyhjä let

solita-juhohaa commented on etp-backend/src/test/clj/solita/etp/service/aineisto_test.clj:
```diff
@@ -74,3 +74,21 @@
+      (t/is (solita.etp.service.file/file-exists? ts/*aws-s3-client* "/aineistot/3/energiatodistukset.csv")))
+    #_(t/testing "Adding some new energiatodistus has an effect on the aineistot"
+      ;;TODO: Could get object size via GetObjectAttributes request parameter `x-amz-object-attributes ObjectSize`
+      ;;      and look that the size increases when energiatodistus is added and aineistot generated again.
```
Katso esimerkkiä palveluväylän testeistä siitä kuinka saa signattuja energiatodistuksia

Tai energiatodisutushausta

solita-juhohaa commented on etp-backend/src/main/clj/solita/etp/service/file.clj:
```diff
@@ -28,4 +28,53 @@
+           (upload-part (byte-array (repeatedly (* 2 1024 1024) #(rand-int 256))))))"
+
+  [aws-s3-client key upload-parts-fn]
+  ;; TODO: Should a failed upload be aborted? Now the cleaning of the failed
```
Muuta Noteksi

solita-juhohaa replied:
Siirretty docstringiin

solita-juhohaa committed:
AE-2036: Test that new energiatodistus shows up in updated aineisto

solita-juhohaa committed:
AE-2036: Rename aineisto-reducible-qyery! -> aineisto-reducible-query

Reading from a database is considered safe in STM transactions.

Juholei commented on etp-backend/src/main/clj/solita/etp/api/aineisto.clj:
```diff
@@ -31,25 +22,20 @@
+                                request
+                                (merge (api-response/csv-response-headers "energiatodistukset.csv" false)
+                                       (api-response/async-cache-headers 86400))
+                                (aineisto-service/aineisto-reducible-query db whoami aineisto-id))
```
Onko tämä route se nykytoteutus vai mikä tämä on?

muep replied:
Tämä on se nykyinen toteutus. Minusta se kannattaisi vielä toistaiseksi säilyttää, on kiva vähintäänkin paikallisen kehityksen kanssa ja signed urlin allekirjoituksen oikeellisuuden tarkistamiseen.

Jos etp-frontin webpacki saadaan opetettua tarjoilemaan miniosta tämän endpointin polku, niin sitten tälle ei välttämättä jäisi mitään käyttöä.

solita-juhohaa replied:
Osittain. Tapaukseen liittyy myös tuo tämän alla oleva `external-routes`. Ymmärtääkseni tekee tuon vastauksen kerran ja cachee sen 24 tunniksi.

Kun CloudFront laitetaan tarjoilemaan nuo csv:t suoraan tuolta S3:sta, niin nämä pitäisi varmaankin poistaa 🤔 

Tämä muutos tehty siis siksi, että siirretty näitä funktioita servicen puolelle. Tässä siis tuo, mitä `result` oli, muutettu tulemaan tuon `aineisto-service/aineisto-reducible-query` kautta.

solita-juhohaa replied:
Näköjään kirjoittelin tähän vaikka Joonas olikin jo kirjoittanut :D

Juholei commented on etp-backend/src/main/clj/solita/etp/api/aineisto.clj:
```diff
@@ -89,3 +75,18 @@
+             :middleware [[security/wrap-whoami-for-internal-aineisto-api]]
+             :responses  {200 {:body nil}}
+             :handler    (fn [{:keys [db whoami aws-s3-client]}]
+                           (concurrent/run-background
```
Ei sinällään vaikuta lopputulokseen, mutta koska run-background vaikuttaisi palauttavan heti nillin, niin voisi tuo r/response olla vain tämän run-background-kutsun ympärillä.

Juholei commented on etp-backend/src/main/clj/solita/etp/api/aineisto.clj:
```diff
@@ -89,3 +75,18 @@
+  [["/aineistot"
+    ["/update"
+     {:post {:summary    "Päivitä kaikkien aineistojen CSV-tiedostot S3:ssa."
+             :middleware [[security/wrap-whoami-for-internal-aineisto-api]]
```
Tässä taisi olla aiemmin myös se oikeustarkistus, mutta onko tämä middlewarekin nyt turha?

solita-juhohaa replied:
En ole varma. Tähän voisi siis varmaan kovakoodata tuon whoamin vain olemaan { :rooli -1 }.

`whoami` menee tuonne funktiolle `solita.etp.service.energiatodistus-search/sql-query`, jossa haarautuu siten, että

1: katsotaan, että onko se nil. Jos on, käytetään public-search-schemaa, jos ei, niin käytetään priate-search-schemaa. Joten täytyy siis ainakin olla jotain.

2: Katsotaan, että on pääkäyttäjä ja voidaan antaa kaikki energiatodistukset, jotka ovat tiloissa 2, 3 tai 4.

Id:tä ei siis nähdäkseni käytetä kyllä nyt missään. Mutta jos jotain logitusta ollaan joskus parantamassa, niin onhan se ehkä hyvä olla? 🤔

Juholei commented on etp-backend/src/main/clj/solita/etp/service/aineisto.clj:
```diff
@@ -65,3 +67,50 @@
         (set-access! db kayttaja-id aineisto))
       1)))
+
+(defn extract-byte-array-and-reset! [buf]
```
Tässä on varmaan aina tiedossa mitä luokkaa tuo buf on? Voisi näin Java-interopia käytettäessä laittaa tälle parametrille type hintin siitä luokasta, niin sitten editoritkin osaa näyttää noiden Java-metodien kutsuista lisätietoa. Sinällään toteutukseen sillä ei ole tällä hetkellä vaikutusta, koska reflektiosta ei tule varoitusta (reflektion kanssa voisi olla heikompi suorituskyky verrattuna siihen, että type hintistä voidaan suoraan päätellä käytettävä luokka), mutta näin kehittäjäkokemuksen ja ymmärrettävyyden kannalta voisi olla parempi lisätä tähän tuo type hint.

Juholei commented on etp-backend/src/main/clj/solita/etp/service/aineisto.clj:
```diff
@@ -65,3 +67,50 @@
+    (.clear buf)
+    arr))
+
+(defn not-nil-aineisto-source! [x]
```
Minä ottaisin huutomerkin pois tästä nimestä, mutta jos tällaista käytäntöä on ollut laajemmalti meidän codebasessa niin sitten voinee antaa olla. Mutta ei ole kuitenkaan mistään yleisestä käytännöstä exceptioneiden kanssa kyse.

Juholei commented on etp-backend/src/main/clj/solita/etp/service/aineisto.clj:
```diff
@@ -65,3 +67,50 @@
+   :anonymized-set energiatodistus-csv/energiatodistukset-anonymized-csv})
+
+(defn aineisto-reducible-query [db whoami aineisto-id]
+  (as-> aineisto-id val
```
Hyvin käytetty tätä as-> makroa tässä

Juholei commented on etp-backend/src/main/clj/solita/etp/service/aineisto.clj:
```diff
@@ -65,3 +67,50 @@
+        ;; This part is used to store rows until it reaches 5MB which
+        ;; is the minimum requirement by `upload-part-fn`.
+        current-part (java.nio.ByteBuffer/allocate (* 8 1024 1024))
+        upload-parts-fn (fn [upload-part-fn] (csv-reducible-query (fn [row]
```
Rivittäisin tätä niin, että ensimmäisen funktion parametrien jälkeen tulisi rivinvaihto, niin ei mene tuo koodi noin paljon kokonaisuudessaan oikealle.

Juholei commented on etp-backend/src/main/clj/solita/etp/service/aineisto.clj:
```diff
@@ -65,3 +67,50 @@
+        key (str "/aineistot/" aineisto-id "/energiatodistukset.csv")
+        ;; This part is used to store rows until it reaches 5MB which
+        ;; is the minimum requirement by `upload-part-fn`.
+        current-part (java.nio.ByteBuffer/allocate (* 8 1024 1024))
```
Importaa käytetty luokka

Juholei commented on etp-backend/src/main/clj/solita/etp/service/aineisto.clj:
```diff
@@ -65,3 +67,50 @@
+        current-part (java.nio.ByteBuffer/allocate (* 8 1024 1024))
+        upload-parts-fn (fn [upload-part-fn] (csv-reducible-query (fn [row]
+                                                                    (let [row-bytes (.getBytes row (StandardCharsets/UTF_8))]
+                                                                      (.put current-part row-bytes)
```
Tästä näyttää tulevan reflektiovaroitus, joten type hint jonnekin auttaisi varmaankin asiaa. Ehkä tuohon rowiin riittäisi, jolloin ehkä tiedetään mitä sen .getBytes palauttaisi jolloin olisi ehkä mahdollista tietää, mitä tyyppiä putin parametri on. Tai sitten vaatii useampaan paikkaan

Juholei commented on etp-backend/src/main/clj/solita/etp/service/aineisto.clj:
```diff
@@ -65,3 +67,50 @@
+                          ;;The last part needs to be uploaded separately (unless the size was a multiple of 5MB)
+                          (when (not= 0 (.position current-part))
+                            (upload-part-fn (extract-byte-array-and-reset! current-part))))]
+    (solita.etp.service.file/upsert-file-in-parts aws-s3-client key upload-parts-fn)
```
Requiraa tämä namespace

Juholei commented on etp-backend/src/main/clj/solita/etp/service/file.clj:
```diff
@@ -28,4 +28,53 @@
+          (throw e))))))
+
+(defn upsert-file-in-parts
+  "Uploads a file to S3 in parts.
```
Hyvä docstring 💯

Juholei commented on etp-backend/src/main/clj/solita/etp/service/file.clj:
```diff
@@ -28,4 +28,53 @@
+    \"foo/bar/baz.txt\"
+    (fn [upload-part]
+           (upload-part (byte-array (repeatedly (* 5 1024 1024) #(rand-int 256))))
+           (upload-part (byte-array (repeatedly (* 2 1024 1024) #(rand-int 256))))))"
```
Onko tämä tarkoituksella kahdesti tässä esimerkissä?

solita-juhohaa replied:
Joo. Lisäsin kommentit selventämään esimerkkiä.

Juholei commented on etp-backend/src/main/clj/solita/etp/service/file.clj:
```diff
@@ -28,4 +28,53 @@
+  [aws-s3-client key upload-parts-fn]
+  (let [{:keys [UploadId]} (aws/create-multipart-upload aws-s3-client key)
+        uploaded-parts-vec (atom [])
+        get-next-part-number (let [current-part-number (atom 0)]
```
Voisiko tämä current-part-number olla ulommassa letissä määriteltynä vai tarvitaanko tässä sulkeumaa?

solita-juhohaa replied:
Voisi olla, mutta ideana tässä oli kätkeä se tuonne funktion sisään, jotta voi käyttää sitä vain tuon `get-next-part-number` kautta.

Juholei commented on etp-backend/src/main/clj/solita/etp/service/file.clj:
```diff
@@ -28,4 +28,53 @@
+                                                                :part-number part-number
+                                                                :upload-id   UploadId
+                                                                :body        content-byte-array})]
+                           (swap! uploaded-parts-vec conj [(str part-number) ETag])))]
```
Voisiko part-numberin ja etagin laittaa jo tässä mappiin, kun niitä tuossa seuraavassa vaiheessa sitten käsitellään niin, että ne laitetaan mapiin?

solita-juhohaa replied:
Joo, hyvä huomio.

Juholei commented on etp-backend/src/main/clj/solita/etp/service/kayttaja.clj:
```diff
@@ -94,4 +94,5 @@
    :laskutus -2
-   :presigned -4})
+   :presigned -4
+   :aineisto -5})
```
Jäikö tämä turhaksi?

Juholei commented on etp-backend/src/test/clj/solita/etp/service/aineisto_test.clj:
```diff
@@ -74,3 +80,84 @@
+        ;; Update aineistot
+        (aineisto/update-aineistot-in-s3! ts/*db* {:id -5 :rooli 2} ts/*aws-s3-client*)
+
+        ;; Aineisto 1
```
Laittaisin näihin kaikki assertioita edeltäviin kommentteihin vielä selityksen, että mitä tässä tarkistetaan.

Juholei commented on etp-backend/src/test/clj/solita/etp/service/aineisto_test.clj:
```diff
@@ -74,3 +80,84 @@
+    (t/is (false? (file/file-exists? ts/*aws-s3-client* "/aineistot/1/energiatodistukset.csv")))
+    (t/is (false? (file/file-exists? ts/*aws-s3-client* "/aineistot/2/energiatodistukset.csv")))
+    (t/is (false? (file/file-exists? ts/*aws-s3-client* "/aineistot/3/energiatodistukset.csv"))))
+  (t/testing "Aineistot exist after generating"
```
Ehkä rivinvaihdot näihin testing-lohkojen väliin, että on selvempää vilkaisulla missä yksi testi loppuu ja seuraava alkaa

muep commented on etp-backend/src/test/clj/solita/etp/service/aineisto_test.clj:
```diff
@@ -74,3 +80,84 @@
+        (t/is (nil? second)))
+
+      ;; Aineisto 3
+      (let [[first second] (get-first-two-energiatodistus-lines-from-aineisto "/aineistot/3/energiatodistukset.csv")]
```
Jos rakennukselle tulee semmoinen käyttötarkoitusluokka jota anonymisointi ei yritä suojata, niin sitten saattaakin käydä niin että sisältyy myös anonymisointiin.

solita-juhohaa replied:
Lisäsin `(assoc-in [:perustiedot :kayttotarkoitus] "RT")`, niin saa sen tuonne anonymisointiinkin.

muep commented on etp-backend/src/main/clj/solita/etp/handler.clj:
```diff
@@ -172,7 +172,8 @@
      (concat (tag "Laskutus API" laskutus-api/routes)
-             (tag "Laatija Internal API" laatija-api/internal-routes))]]
+             (tag "Laatija Internal API" laatija-api/internal-routes)
+             (tag "Aineisto API" aineisto-api/internal-routes))]]
```
Tässä voisi kans olla mainittuna, että on internal

solita-juhohaa replied:
Laitoin, koska olisi muuten ollut ainoa internal route, joka ei ole oman otsikkonsa alla.

muep reviewed (approved):
Muutama pikkujuttu tuolta nousi kommentoitavaa

solita-juhohaa committed:
AE-2036: Update README to include information about the MinIO Console

solita-juhohaa committed:
AE-2036: Add type hints and use imports

solita-juhohaa committed:
AE-2036: Rename not-nil-aineisto-source! -> not-nil-aineisto-source

solita-juhohaa committed:
AE-2036: Fix indentation

solita-juhohaa committed:
AE-2036: Add type hints and use imports

solita-juhohaa committed:
AE-2036: Fix upsert-file-in-parts

- Clarify the example in docstring
- Remove unnecessary intermediate step when creating the uploaded-parts-vec
- Actually return nil

solita-juhohaa committed:
AE-2036: Fix update-aineistot-test

- Improve comments
- Don't rely on the order of the line in the csv
- Add todistukset asap to test that signing them
  is the operation that makes them show in the csv
- Change kayttotarkoitus to "RT" (rivitalo) to
  make something appear in the anonymized aineisto also

solita-juhohaa committed:
AE-2036: Run actions inside of response

solita-juhohaa committed:
AE-2036: Add 'Internal' to Aineisto API's internal routes' tag

solita-juhohaa committed:
AE-2036: Add middleware to add `application_name` to database calls
